server/util/networking/networking.go

```diff
@@ -642,31 +590,92 @@ func (s cleanupStack) Cleanup(ctx context.Context) error {
 	return nil
 }
 
-// ConfigureNATForTapInNamespace configures NAT in the namespace so that
-// outgoing IP packets from the tap device in the namespace are translated to
-// the namespaced veth device IP, and incoming IP packets to the tap device are
-// translated to the VM tap device IP.
+// VMNetwork represents a fully-provisioned VM network, which consists of a net
+// namespace, virtual network interfaces, and associated host configuration.
 //
-// Since the host-side of the veth pair also has NAT configured, this allows the
-// VM to communicate with external networks.
-func ConfigureNATForTapInNamespace(ctx context.Context, vethPair *VethPair, vmIP string) error {
-	if err := runCommand(ctx, namespace(vethPair.netns, "iptables", "--wait", "-t", "nat", "-A", "POSTROUTING", "-o", vethPair.namespacedDevice, "-s", vmIP, "-j", "SNAT", "--to", vethPair.network.NamespacedIP())...); err != nil {
-		return err
+// Deleting a VM network deletes the net namespace as well as all associated
+// resources, and reverts the applied host configuration.
+type VMNetwork struct {
+	netns    *Namespace
+	vethPair *vethPair
+	cleanup  func(ctx context.Context) error
+}
+
+// CreateVMNetwork initializes a network namespace, networking
+// interfaces, and host configuration required for VM networking.
+func CreateVMNetwork(ctx context.Context, tapDeviceName, tapAddr, vmIP string) (_ *VMNetwork, err error) {
+	var cleanupStack cleanupStack
+	defer func() {
+		// If we failed to fully set up the network, make sure to clean up any
+		// resources that were partially set up.
+		if err != nil {
+			_ = cleanupStack.Cleanup(ctx)
+		}
+	}()
+
+	// Create a net namespace.
+	netns, err := createUniqueNetNamespace(ctx)
+	if err != nil {
```

Comment:
Previously we only warned if we got a "file exists" error here - was it intentional to remove that?

Reply:
I removed that code because that error doesn't get logged in practice, which makes sense because net namespace names are generated using a uuid - so it shouldn't be possible for a namespace to already exist when we create it